GH-36155: [C++][Go][Java][FlightRPC] Add support for long-running queries (#36946)
apache/arrow: docs/source/format/Flight.rst

.. _google.protobuf.Timestamp: https://protobuf.dev/reference/protobuf/google.protobuf/#timestamp

Downloading Data by Running a Heavy Query
-----------------------------------------

> A client may need to request a heavy query to download
> data. ``GetFlightInfo`` isn't returned until the requested query is
> completed. It means that the client is blocked. The client can use
> ``PollFlightInfo`` instead of ``GetFlightInfo``:

Reviewer: ```suggestion
A client may need to request a heavy query to download
data. However, ``GetFlightInfo`` doesn't return until the query
completes, so the client is blocked. In this situation, the client
can use ``PollFlightInfo`` instead of ``GetFlightInfo``:
```

   response.

   If the query isn't finished, ``RetryInfo.flight_descriptor`` has a
   ``FlightDescriptor``. The client should use the descriptor (not the
   original ``FlightDescriptor``) to call the next
   ``PollFlightInfo()``.  A server should recognize a
   ``RetryInfo.flight_descriptor`` that is not necessarily the latest
   in case the client misses an update in between.

   If the query is finished, ``RetryInfo.flight_descriptor`` is
   unset.

   ``RetryInfo.info`` is the currently available results so far. It's
   a complete ``FlightInfo`` each time not just the delta between the
   previous and current ``FlightInfo``. A server should only append to
   the endpoints in ``RetryInfo.info`` each time. So the client can
   run ``DoGet(Ticket)`` with the ``Ticket`` in the ``RetryInfo.info``
   even when the query isn't finished yet. ``FlightInfo.ordered`` is
   also valid.

   A server should not respond until the result would be different
   from last time. That way, the client can "long poll" for updates
   without constantly making requests. Clients can set a short timeout
   to avoid blocking calls if desired.

>    ``RetryInfo.progress`` may be set. It represents progress of the
>    query. If it's set, the value must be in ``[0.0, 1.0]``. The value
>    must not be monotonic or nondecreasing. A server may respond by
>    only updating the ``RetryInfo.progress`` value though it shouldn't
>    spam the client with updates.

Reviewer: ```suggestion
   ``RetryInfo.progress`` may be set. It represents progress of the
   query. If it's set, the value must be in ``[0.0, 1.0]``. The value
   is not necessarily monotonic or nondecreasing. A server may respond by
   only updating the ``RetryInfo.progress`` value, though it shouldn't
   spam the client with updates.
```
